### ISSUE TITLE
implement passing absolute stop signal on restricted aspect ( blinking red ) ? 

### ISSUE THREAD

**Issue by 2-8-4** [ecgconnelly/uss8#22]
don't know if and how this was implemented on CTC panels with the present levers and buttons

**Comment by ecgconnelly** [comment]
This is in progress - can't believe I didn't have an issue for it set up 🤦‍♀️

**Comment by 2-8-4** [comment]
No need to rush, Emilia.

Let the bug reports and feature requests come in so you can prioritize and sort them out later.

I’ll go looking for some icons on that JMRI site.
Are your panel icons in your GitHub?

You program USS8 in C language , while JMRI and that failed commercial external dispatcher were programmed in Java.

In theory it would be possible to build a hardware CTC panel even and interface this to Run8 instead of a model railroad. How cool is that?

Keep up your good work, I like it very much.

Edwin Koppers,
Ruurlo,
The Netherlands

Op 4 dec. 2022 om 19:22 heeft Emilia Connelly ***@***.***> het volgende geschreven:

﻿

This is in progress - can't believe I didn't have an issue for it set up 🤦‍♀️

**Comment by ecgconnelly** [comment]
Haha the icons are literally from JMRI, through the magic of free software
and the GPL license.  The icons are in both this repo and JMRI.

Really, TSCAD was java?  That explains a lot.  And yeah, you could
absolutely build a hardware panel - but you know Run8 would change their
track layout the next week.

**Comment by 2-8-4** [comment]
Thought so already about those icons.

I see you’ve removed the Bakersfield Amtrak switch I suggested from your GitHub.
The track layout on the panel is OK with me, only the switch can be reversed, Normal is towards Kern Jct, Reverse towards Amtrak to have a more logical operation. Just add a scribble on the panel to point it out ;-)

Anyway, since I don’t know you intend to replicate the real CTC panel that perhaps once existed or just design an interpretation, the choice is up to the programmer.

Additional features require additional switches, buttons or levers in the columns.

Like silencing annunciator bell, blue flagging tracks, release switch for manual operation, opening locked derails, passing absolute red on restriction, etc.

Lots of work to design and implement in the CTC logic and your program.

For starters it already is very nice and free.

First sort out any bugs and errors.

Edwin Koppers,
Ruurlo,
The Netherlands

Op 5 dec. 2022 om 09:24 heeft Emilia Connelly ***@***.***> het volgende geschreven:

﻿

Haha the icons are literally from JMRI, through the magic of free software
and the GPL license. The icons are in both this repo and JMRI.

Really, TSCAD was java? That explains a lot. And yeah, you could
absolutely build a hardware panel - but you know Run8 would change their
track layout the next week.

**Comment by 2-8-4** [comment]
Hi, Emilia.

You implemented passing an absolute stop signal on restricted aspect by pressing CTRL if I understand correctly?

You ported a model railroad implementation of CTC, and for model railroad control it is hardly necessary to pass an absolute red stop indication. So the JMRI programmers probably didn’t implement this.

Obviously, simulations like Run8 can strictly adhere to prototype practices, and on the real Bakersfield Sub CTC panel where special smaller levers above the normal switch and signal levers. I think specially for the purpose of ‘unlocking’ a control point’s switches and signals for manual switching. Implementing this would require some graphic design, working out the signalling logic, thinking how to implement this in a reasonable fashion for your project.

Perhaps for the future?

But how to differentiate between the absolute stop and restricted after the CTRL press? Will the signal light blink red? What about the direction? What about the switch being manually thrown? The unlocked switch lamps blink, stay dark?

Best read up on prototype practice or find an ATSF USS CTC machine manual at SFHS or somewhere else?

EDwin

**Comment by ecgconnelly** [comment]
Hey Edwin - 

> You implemented passing an absolute stop signal on restricted aspect by pressing CTRL if I understand correctly?

Correct.  This is meant to represent a pull or twist of the code button, as described in this [railroad.net post](https://www.railroad.net/difference-between-a-call-on-signal-and-restricting-t84457.html#p953748).  I am aware that some machines had separate buttons the operator would use to request a call-on but I don't see much of a benefit in adding that.



> You ported a model railroad implementation of CTC

Not really.  USS8 currently uses JMRI graphics, though we do have an individual currently working on much nicer replacements.  JMRI logic is not used and honestly probably wouldn't be that helpful.  



> special smaller levers above the normal switch and signal levers

This is exactly the plan for unlocking switches.  For now, that will probably take the form of copying the fleet switch images to a position near the switch lever.  We'll see how froggy our graphics contributor wants to get.



> But how to differentiate between the absolute stop and restricted after the CTRL press?

I could be wrong, and prototypes may differ, but to my knowledge, the current lamp behavior after requesting a call-on signal is correct for at least some prototypes.  Restricting is not Stop, therefore the machine displays a Proceed indication on the panel.



> What about the switch being manually thrown? The unlocked switch lamps blink, stay dark?

Probably go dark.  I seem to recall reading something to this effect but this is easily changed if information to the contrary surfaces.



> Best read up on prototype practice or find an ATSF USS CTC machine manual at SFHS or somewhere else?

I'd love to see such a manual.  I've looked and although there is more information now than I was able to find circa 2015 (see the veritable trove [here](http://rrsignalpix.com/documents.php)) I have been unable to find an actual manual.  Plenty of catalogs, articles, and individual machine drawings but not much beyond that.  In any case the research aspect isn't all that fun for me...
